File: bindings/lua/rns/identity.lua
-- SPDX-License-Identifier: Apache-2.0
-- Copyright (c) 2024-2026 Quad4.io

local ffi_mod = require("rns.ffi")
local errors = require("rns.errors")
local ffi = ffi_mod.ffi
local lib = ffi_mod.lib

local Identity = {}
Identity.__index = Identity

local Handle = ffi.typeof("struct { uint64_t h; }")

local function wrap(handle)
	local box = ffi.new(Handle)
	box.h = handle
	local self = setmetatable({ _box = box }, Identity)
	ffi.gc(box, function(b)
		if b.h ~= 0 then
			lib.rns_identity_destroy(b.h)
			b.h = 0
		end
	end)
	return self
end

function Identity.generate()
	local h = lib.rns_identity_generate()
	if h == 0 then
		errors.raise(errors.Error.INTERNAL)
	end
	return wrap(h)
end

function Identity.load(path)
	if not path or path == "" then
		errors.raise(errors.Error.INVALID_ARG)
	end
	local h = lib.rns_identity_load(path)
	if h == 0 then
		errors.raise(errors.Error.IO)
	end
	return wrap(h)
end

function Identity.from_public_key(pub)
	local buf, n = ffi_mod.to_uint8(pub)
	if n == 0 then
		errors.raise(errors.Error.INVALID_ARG)
	end
	local h = lib.rns_identity_from_public_key(buf, n)
	if h == 0 then
		errors.raise(errors.Error.INVALID_ARG)
	end
	return wrap(h)
end

function Identity:save(path)
	if not path or path == "" then
		errors.raise(errors.Error.INVALID_ARG)
	end
	errors.map_code(lib.rns_identity_save(self._box.h, path))
end

function Identity:hash_hex()
	local buf = ffi.new("char[64]")
	local written = ffi.new("size_t[1]", 0)
	errors.map_code(lib.rns_identity_hash(self._box.h, buf, 64, written))
	return ffi.string(buf, written[0])
end

function Identity:hash_bytes()
	local out = ffi.new("uint8_t[?]", ffi_mod.HASH_LEN)
	local written = ffi.new("size_t[1]", 0)
	errors.map_code(lib.rns_identity_hash_bytes(self._box.h, out, ffi_mod.HASH_LEN, written))
	if written[0] ~= ffi_mod.HASH_LEN then
		errors.raise(errors.Error.TRUNCATED)
	end
	return ffi.string(out, written[0])
end

function Identity:public_key()
	local out = ffi.new("uint8_t[64]")
	local written = ffi.new("size_t[1]", 0)
	errors.map_code(lib.rns_identity_public_key(self._box.h, out, 64, written))
	return ffi.string(out, written[0])
end

function Identity:sign(data)
	local buf, n = ffi_mod.to_uint8(data)
	local sig = ffi.new("uint8_t[64]")
	local written = ffi.new("size_t[1]", 0)
	errors.map_code(lib.rns_identity_sign(self._box.h, buf, n, sig, 64, written))
	return ffi.string(sig, written[0])
end

function Identity:verify(data, signature)
	local dbuf, dn = ffi_mod.to_uint8(data)
	local sbuf, sn = ffi_mod.to_uint8(signature)
	errors.map_code(lib.rns_identity_verify(self._box.h, dbuf, dn, sbuf, sn))
end

function Identity:handle()
	return self._box.h
end

function Identity:close()
	if self._box.h ~= 0 then
		lib.rns_identity_destroy(self._box.h)
		self._box.h = 0
	end
end

return Identity

File: bindings/java/src/io/quad4/rns/Identity.java
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

package io.quad4.rns;

import com.sun.jna.ptr.LongByReference;
import io.quad4.rns.ffi.RnsLibrary;

/** Opaque identity handle. */
public final class Identity implements AutoCloseable {
    private long handle;

    Identity(long handle) {
        this.handle = handle;
    }

    public static Identity generate() {
        long h = RnsLibrary.INSTANCE.rns_identity_generate();
        if (h == 0) {
            throw new RnsException(RnsException.INTERNAL);
        }
        return new Identity(h);
    }

    public static Identity load(String path) {
        if (path == null || path.isEmpty()) {
            throw new RnsException(RnsException.INVALID_ARG);
        }
        long h = RnsLibrary.INSTANCE.rns_identity_load(path);
        if (h == 0) {
            throw new RnsException(RnsException.IO);
        }
        return new Identity(h);
    }

    public static Identity fromPublicKey(byte[] pub) {
        if (pub == null || pub.length == 0) {
            throw new RnsException(RnsException.INVALID_ARG);
        }
        long h = RnsLibrary.INSTANCE.rns_identity_from_public_key(pub, pub.length);
        if (h == 0) {
            throw new RnsException(RnsException.INVALID_ARG);
        }
        return new Identity(h);
    }

    public long handle() {
        return handle;
    }

    public void save(String path) {
        if (path == null || path.isEmpty()) {
            throw new RnsException(RnsException.INVALID_ARG);
        }
        Rns.check(RnsLibrary.INSTANCE.rns_identity_save(handle, path));
    }

    public String hashHex() {
        byte[] buf = new byte[64];
        LongByReference written = new LongByReference();
        Rns.check(RnsLibrary.INSTANCE.rns_identity_hash(handle, buf, buf.length, written));
        return new String(buf, 0, (int) written.getValue());
    }

    public byte[] hashBytes() {
        byte[] out = new byte[Rns.HASH_LEN];
        LongByReference written = new LongByReference();
        Rns.check(RnsLibrary.INSTANCE.rns_identity_hash_bytes(handle, out, out.length, written));
        if (written.getValue() != Rns.HASH_LEN) {
            throw new RnsException(RnsException.TRUNCATED);
        }
        return out;
    }

    public byte[] publicKey() {
        byte[] out = new byte[64];
        LongByReference written = new LongByReference();
        Rns.check(RnsLibrary.INSTANCE.rns_identity_public_key(handle, out, out.length, written));
        byte[] result = new byte[(int) written.getValue()];
        System.arraycopy(out, 0, result, 0, result.length);
        return result;
    }

    public byte[] sign(byte[] data) {
        byte[] payload = data == null ? new byte[0] : data;
        byte[] sig = new byte[64];
        LongByReference written = new LongByReference();
        Rns.check(RnsLibrary.INSTANCE.rns_identity_sign(handle, payload, payload.length, sig, sig.length, written));
        byte[] result = new byte[(int) written.getValue()];
        System.arraycopy(sig, 0, result, 0, result.length);
        return result;
    }

    public void verify(byte[] data, byte[] signature) {
        byte[] payload = data == null ? new byte[0] : data;
        byte[] sig = signature == null ? new byte[0] : signature;
        Rns.check(RnsLibrary.INSTANCE.rns_identity_verify(handle, payload, payload.length, sig, sig.length));
    }

    @Override
    public void close() {
        if (handle != 0) {
            RnsLibrary.INSTANCE.rns_identity_destroy(handle);
            handle = 0;
        }
    }
}
